### PR TITLE
Distributed channel - watch config-kafka like consolidated channel

### PR DESCRIPTION
**Pull request by aliok**
Addresses controller part of #455.

I tried to do things in small steps in separate PRs.

## Proposed Changes

- Watch the configmap in the controller (already existing)
- When the configmap is changed, do a global resync of the KafkaChannel CRs so that they're reconciled

- Bonus: copy over some good stuff from distributed channel to consolidated channel (some nil checks etc. I found missing in consolidated channel)

I tested my changes with https://github.com/aliok/kafka-ch-reload-tests/tree/distr-channel (beware the `distr-channel` branch). There are 2 pods:
- sender sends messages for 2 mins
- receiver receives messages for 3 mins

I haven't seen any message lost (sent to Kafka but not received by the receiver) or non-posted (couldn't sent by the sender to Kafka).

FYI, this the plan I have in my mind:
- Do a global resync of CRs when configmap changes
- Remove configmap watching in dispatcher and receiver and mount the configmap to these
- Controller to trigger a rolling update of dispatcher and receiver deployments when configmap changes (similar to https://github.com/knative-sandbox/eventing-kafka/pull/514)

With these done, the behaviour and code will look very very similar on both channel implementations when it comes to handling the changes in `config-kafka`.

**Release Note**

```release-note

```

**Docs**

### CODE REVIEW

**Comment by aliok** [comment]
cc @eric-sap mind having a look?

**Comment by eric-sap** [comment]
From what I can tell, this should do "what it did before" in the distributed channel when a the configmap changes, but also do a resync (re-reconcile).  Right now that won't do anything but could eventually use the hash differences to do rollout-restart-style updates of the deployments.   If that's an accurate summary then it looks good; thanks!

**Comment by aliok** [comment]
> Right now that won't do anything but could eventually use the hash differences to do rollout-restart-style updates of the deployments.

Yes, exactly.

Thanks for the review.

Let me ping @devguyio @slinkydeveloper for lgtm/approve.

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go [hunk 1]
```diff
@@ -568,10 +568,22 @@ func (r *Reconciler) deleteTopic(ctx context.Context, channel *v1beta1.KafkaChan
...
+	if configMap == nil {
+		logger.Warn("Nil ConfigMap passed to configMapObserver; ignoring")
+		return
+	}
+
 	logger.Info("Reloading Kafka configuration")
 	kafkaConfig, err := utils.GetKafkaConfig(configMap.Data)
 	if err != nil {
 		logger.Errorw("Error reading Kafka configuration", zap.Error(err))
+		return
```
Comment (slinkydeveloper) [inline pkg/channel/consolidated/reconciler/controller/kafkachannel.go:586]: Should you return an error here?

Comment (aliok) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:586]: I will give it a try

Comment (aliok) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:586]: This function is not called within the reconciliation loop. It is called when config-kafka is changed.
A global resync is done later.

Comment (slinkydeveloper) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:586]: ah ok gotcha!

**Review by slinkydeveloper — APPROVED** [review]
/lgtm

**Pull request merged**